### PR TITLE
[ARO] Add enhancements as well as route table checking permissions

### PR DESCRIPTION
**Pull request by kwoodson**
**Description**  

The updates here are the latest development features to be added to Azure Red Hat OpenShift az cli client.
- enhanced permissions validation
- logging improvements
- type checking

**Testing Guide**  

None
These are teh enhancements that we use for our Azure cli in the github.com/Azure/ARO-RP repository.

**History Notes**  

None
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by kwoodson** [comment]
@jim-minter @mjudeikis @ehashman  Updates to ARO CLI extension.

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py [hunk 1]
```diff
@@ -29,17 +34,63 @@ def assign_contributor_to_vnet(cli_ctx, vnet, object_id, label):
...
-                assignment.principal_id.lower() == object_id.lower():
-            return
+    if has_assignment(auth_client.role_assignments.list_for_scope(vnet), role_definition_id, object_id):
+        return
 
-    # use object_id as a namespace and vnet and label as a salt for generating UUID to
-    # make it stable across calls
-    role_uuid = uuid.uuid5(uuid.UUID(object_id), ','.join(vnet + label))
+    # generate random uuid for role assignment
+    role_uuid = _gen_uuid()
```
Comment (kwoodson) [inline src/azure-cli/azure/cli/command_modules/aro/_rbac.py:41]: Wasn't sure if we should keep the previous method of generating a role_uuid.
```
role_uuid = uuid.uuid5(uuid.UUID(object_id), ','.join(vnet + label))
```

File: src/azure-cli/azure/cli/command_modules/aro/custom.py [hunk 2]
```diff
@@ -49,7 +53,7 @@ def aro_create(cmd,  # pylint: disable=too-many-locals
 
     aad = AADManager(cmd.cli_ctx)
     if client_id is None:
-        app, client_secret = aad.create_application('aro-%s' % random_id)
+        app, client_secret = aad.create_application(cluster_resource_group or 'aro-' + random_id)
```
Comment (kwoodson) [inline src/azure-cli/azure/cli/command_modules/aro/custom.py:56]: Also wasn't sure about this `cluster_resource_group` here.  Should we remove this and leave this as-is?

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py (hunk 1 above)
Comment (jim-minter) [reply src/azure-cli/azure/cli/command_modules/aro/_rbac.py:41]: no - all references to uuid5 should be removed.

File: src/azure-cli/azure/cli/command_modules/aro/custom.py (hunk 2 above)
Comment (jim-minter) [reply src/azure-cli/azure/cli/command_modules/aro/custom.py:56]: +1 on this change

File: src/azure-cli/azure/cli/command_modules/aro/custom.py [hunk 3]
```diff
@@ -41,6 +44,7 @@ def aro_create(cmd,  # pylint: disable=too-many-locals
                ingress_visibility=None,
                tags=None,
                no_wait=False):
+
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply src/azure-cli/azure/cli/command_modules/aro/custom.py:47]: Yes, I accidentally clipped this when I saw the development environment check.

**Comment by kwoodson** [comment]
@jim-minter, updated the PR to include the requested changes.

I have fixed the spacing in the `_help.py` file.  It looks like there are still some differences like the `OpenShift` being inserted into the different help sections.  I'd recommend that the next step would be to port these changes locally so that we have fewer differences in the future.  This would be a good time to do so.

File: src/azure-cli/azure/cli/command_modules/aro/_aad.py [hunk 4]
```diff
@@ -56,6 +61,15 @@ def get_service_principal(self, app_id):
...
+        while True:
+            try:
+                return self.client.service_principals.create(
+                    ServicePrincipalCreateParameters(app_id=app_id))
+            except GraphErrorException as ex:
+                if retries >= max_retries:
+                    raise
+                retries += 1
+                logger.warning("%s; retry %d of %d", ex, retries, max_retries)
+                time.sleep(10)
```
Comment (ehashman) [inline src/azure-cli/azure/cli/command_modules/aro/_aad.py:75]: This is very go-esque as written, I'd suggest a refactor to

```suggestion
        max_retries = 3
        for i in range(max_retries):
            try:
                return self.client.service_principals.create(
                    ServicePrincipalCreateParameters(app_id=app_id))
            except GraphErrorException as ex:
                logger.warning("%s; retry %d of %d", ex, i, max_retries)
                time.sleep(10)
```

File: src/azure-cli/azure/cli/command_modules/aro/_help.py [hunk 5]
```diff
@@ -7,66 +7,66 @@
 
 
 helps['aro'] = """
-    type: group
-    short-summary: Manage Azure Red Hat OpenShift clusters.
+  type: group
+  short-summary: Manage Azure Red Hat OpenShift clusters.
```
Comment (ehashman) [inline src/azure-cli/azure/cli/command_modules/aro/_help.py:11]: Are these changes all whitespace? Are they needed or can we exclude them from the patch?

File: src/azure-cli/azure/cli/command_modules/aro/_aad.py (hunk 4 above)
Comment (ehashman) [reply src/azure-cli/azure/cli/command_modules/aro/_aad.py:75]: (Note that the quick refactor doesn't include that final raise, it just gives up -- would need to follow up with an appropriate exception)

**Comment by yonzhan** [comment]
ARO

File: src/azure-cli/azure/cli/command_modules/aro/_help.py (hunk 5 above)
Comment (kwoodson) [reply src/azure-cli/azure/cli/command_modules/aro/_help.py:11]: | Please can you fix the whitespace diffs in _help.py via this PR as well?

@ehashman, Yes, they are.  @jim-minter requested above that we align these with our `_help.py`.

**Comment by ehashman** [comment]
@kwoodson from the CI test failures, it looks like you might need to rerecord the tests as part of this PR

File: src/azure-cli/azure/cli/command_modules/aro/_aad.py (hunk 4 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/_aad.py:75]: These refactor should be done in our repository, let's not diff this even more.

**Comment by haroldrandom** [comment]
@kwoodson Could you solve the CI failure first?

**Comment by kwoodson** [comment]
@haroldrandom, I'll look into the CI failures.

**Comment by kwoodson** [comment]
The command that enabled updating the current tests:
```
. env/bin/activate
azdev test aro --live
```

**Comment by kwoodson** [comment]
@haroldrandom Were these the files required?  I see that it is still failing.
```
[gw6] [ 11%] PASSED src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_helpers.py::TestGenerateRandomIdHelper::test_random_id_length 
[gw7] [ 22%] PASSED src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_helpers.py::TestGenerateRandomIdHelper::test_random_id_is_alpha_num 
[gw8] [ 33%] PASSED src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_helpers.py::TestGenerateRandomIdHelper::test_random_id_starts_with_letter 
[gw4] [ 44%] PASSED src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py::AroScenarioTests::test_aro_show 
[gw2] [ 55%] PASSED src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py::AroScenarioTests::test_aro_list 
[gw5] [ 66%] PASSED src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py::AroScenarioTests::test_aro_update 
[gw3] [ 77%] PASSED src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py::AroScenarioTests::test_aro_list_credentials 
[gw0] [ 88%] PASSED src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py::AroScenarioTests::test_aro_create 
[gw1] [100%] PASSED src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py::AroScenarioTests::test_aro_delete 

======== 9 passed, 1 warning in 3051.90s (0:50:51) ========
```

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py (hunk 1 above)
Comment (petrkotas) [reply src/azure-cli/azure/cli/command_modules/aro/_rbac.py:41]: Yes use the _gen_uuid() is it later used in test recording as it can be mocked.

Comment (kwoodson) [reply src/azure-cli/azure/cli/command_modules/aro/_rbac.py:41]: @petrkotas Thanks!  I've added the mock and rerunning with `--live` to rerecord.  Will update when done.

**Comment by ehashman** [comment]
I believe CI is failing because #13086 has not yet landed. UUID generation does not look stable from the failure results.

File: src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py [hunk 6]
```diff
@@ -45,13 +46,14 @@ def test_aro_create(self, resource_group):
...
+        with mock.patch('azure.cli.command_modules.aro._rbac._gen_uuid', side_effect=self.create_guid):
+            self.cmd('aro create -g {rg} -n {name} --master-subnet {master_subnet_resource} --worker-subnet {worker_subnet_resource} --subscription {subscription} --tags test=create', checks=[
+                self.check('tags.test', 'create'),
+                self.check('name', '{name}'),
+                self.check('masterProfile.subnetId', '{master_subnet_resource}'),
+                self.check('workerProfiles[0].subnetId', '{worker_subnet_resource}'),
+                self.check('provisioningState', 'Succeeded')
+            ])
 
     @AllowLargeResponse()
```
Comment (petrkotas) [inline src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py:58]: I know, I had to remove this line to make it work in my PR. https://github.com/Azure/azure-cli/pull/13086/files#diff-36af8ad76e1e927f0a66611dfd361a53L56

Comment (petrkotas) [reply src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py:58]: I believe it will help here

Comment (kwoodson) [reply src/azure-cli/azure/cli/command_modules/aro/tests/latest/test_aro_commands.py]: Added this as well.  Let's see what happens.

File: src/azure-cli/azure/cli/command_modules/aro/_help.py [hunk 7]
```diff
@@ -7,66 +7,66 @@
...
-      - name: Create an OpenShift cluster with 5 compute nodes and Red Hat Pull Secret
-        text: az aro create --resource-group MyResourceGroup --name MyCluster --vnet MyVnet --master-subnet MyMasterSubnet --worker-subnet MyWorkerSubnet --worker-count 5 --pull-secret @pullsecret.txt
-      - name: Create a Private OpenShift cluster
-        text: az aro create --resource-group MyResourceGroup --name MyCluster --vnet MyVnet --master-subnet MyMasterSubnet --worker-subnet MyWorkerSubnet --apiserver-visibility Private --ingress-visibility Private
+  type: command
+  short-summary: Create a cluster.
+  examples:
+  - name: Create an OpenShift cluster
+    text: az aro create --resource-group MyResourceGroup --name MyCluster  --vnet MyVnet --master-subnet MyMasterSubnet --worker-subnet MyWorkerSubnet
+  - name: Create an OpenShift cluster with 5 compute nodes and Red Hat Pull Secret
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/_help.py:20]: @kwoodson there are still diffs between this and the ARO-RP version.  The ARO-RP version is the one that is correct.  Please can you update this PR accordingly?

File: src/azure-cli/azure/cli/command_modules/aro/_params.py [hunk 8]
```diff
@@ -64,9 +64,11 @@ def load_arguments(self, _):
         c.argument('worker_vm_size',
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/_params.py:64]: There is a diff in this file around `get_location_type(self.cli_ctx)` between here and ARO-RP.  Not sure which is right.  Please can you check?

File: src/azure-cli/azure/cli/command_modules/aro/_params.py [hunk 9]
```diff
@@ -64,9 +64,11 @@ def load_arguments(self, _):
...
                    help='Size of worker VMs.')
         c.argument('worker_vm_disk_size_gb',
+                   type=int,
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/_params.py:67]: there are corresponding diffs which need to be fixed up in _validators.py - please check

**Comment by jim-minter** [comment]
@kwoodson @ehashman the diff I am trying to minimise is `diff -uNr -x '*.pyc' -x vendored_sdks -x tests src/azure-cli/azure/cli/command_modules/aro ~/go/src/github.com/Azure/ARO-RP/python/az/aro/azext_aro/ | less`

**Comment by ehashman** [comment]
Thanks for the helpful context and feedback @jim-minter 

@petrkotas's PR has now merged, so this will need to be rebased. I suggest we close this PR and open a new one, WDYT @kwoodson ?

**Comment by kwoodson** [comment]
@ehashman sgtm

**Comment by jim-minter** [comment]
@kwoodson is there a new PR for this?

**Comment by kwoodson** [comment]
@jim-minter I have added the new PR here https://github.com/Azure/azure-cli/pull/14533

**Pull request closed without merging**